src/onewire.rs

```diff
@@ -0,0 +1,135 @@
+//!
+//!
+//! This implementation currently supports the one-wire API from the new (v5) esp-idf API.
+//!
+//! The pin this peripheral is attached to must be
+//! externally pulled-up with a 4.7kOhm resistor.
+//!
+//! See the `examples/` folder of this repository for more.
+
+use core::marker::PhantomData;
+use core::ptr;
+
+use esp_idf_sys::*;
+
+use crate::peripheral::Peripheral;
+
+#[derive(Debug)]
+pub struct Device<'b> {
+    _device: onewire_device_t,
+    // _bus: &'b BusDriver<'b>, // not sure how I would hold a reference to this aside from the handle
+    _p: PhantomData<&'b ()>, // holds the lifetime since the device is linked to the lifetime of the bus
+}
+
+impl<'a> Device<'a> {
+    fn new(device_handle: onewire_device_t) -> Self {
+        Self {
+            _device: device_handle,
+            _p: PhantomData,
+        }
+    }
+
+    /// get the handle of the bus the device is attached to
+    pub fn bus(&self) -> onewire_bus_handle_t {
+        self._device.bus
+    }
+
+    /// get the device address
+    pub fn address(&self) -> u64 {
+        self._device.address
+    }
+}
+
+/// wrapper around the device iterator to search for available devices on the bus
+pub struct DeviceSearch<'b> {
+    _search: onewire_device_iter_handle_t,
+    _p: PhantomData<&'b ()>, // holds the lifetime since the search is linked to the lifetime of the bus
+}
+
+impl<'b> DeviceSearch<'b> {
+    fn new(bus: &mut OneWireBusDriver) -> Result<Self, EspError> {
+        let mut my_iter: onewire_device_iter_handle_t = ptr::null_mut();
+
+        esp!(unsafe { onewire_new_device_iter(bus._bus, &mut my_iter) })?;
+
+        Ok(Self {
+            _search: my_iter,
+            _p: PhantomData,
+        })
+    }
+
+    /// Search for the next device on the bus and yield it.
+    fn next_device(&mut self) -> Result<Device<'b>, EspError> {
+        let mut next_onewire_device = onewire_device_t::default();
+        esp!(unsafe { onewire_device_iter_get_next(self._search, &mut next_onewire_device) })?;
+
+        Ok(Device::new(next_onewire_device))
+    }
+}
+
+impl<'b> Iterator for DeviceSearch<'b> {
+    type Item = Device<'b>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if let Ok(dev) = self.next_device() {
+            Some(dev)
+        } else {
+            None
+        }
+    }
+}
+
+impl<'b> Drop for DeviceSearch<'b> {
+    fn drop(&mut self) {
+        esp!(unsafe { onewire_del_device_iter(self._search) }).unwrap();
+    }
+}
+
+pub struct OneWireBusDriver<'d> {
+    _bus: onewire_bus_handle_t,
+    _p: PhantomData<&'d ()>,
+}
+
+impl<'d> OneWireBusDriver<'d> {
+    /// Create a new One Wire driver on the allocated pin.
+    ///
+    /// The pin will be used as an open drain output.
+    pub fn new(
+        pin: impl Peripheral<P = impl crate::gpio::InputPin + crate::gpio::OutputPin> + 'd,
```

Comment:
Since internally the ESP IDF onewire driver uses the new RMT driver (a single channel for TX/RX) (plus the pin thst you already track) you need to somehow track the usage of the RMT channel peripheral, as it is still a peripheral - just like the pin which you track already.

If you don't track it, nothing major would happen because the old and the new RMT driver cannot co-exist, yet would still be good to follow the pattern established with the other drivers.

Now, some of the "new" (ESP IDF 5+) C drivers no longer take a peripheral "id"/"number". For example, the old RMT driver used to take a channel number, but the new C RMT driver doesn't as it internally manages what channel to be used. If you try to instantiate more driver instances than hardware channels you have available, it would likely return a runtime `EspError`. So in a way - with the new RMT C driver you don't really know which concrete hardware channel peripheral will be used by that particular C driver instance.

With that said, _nothing stops us from continuing the old way and passing to the driver a RMT "peripheral"_ (note that you don't need "new" RMT channel peripherals for the new C RMT driver, because the peripherals are ghost structures which however model **actual** hardware on the chip and not higher-level C driver concepts).

So what I suggest is to just enhance the `new` constructor by passing an `impl ... Channel ...` peripheral that you can copy from the old RMT driver wrapper. While there is no warranty that the C driver will use _exactly_ the same channel that you pass to it - we don't care, as the channels are indistinguishable from each other. We just use - with the new C driver - the channels as a way of glorified "reference counting" of sorts.

Reply:
Yeah I had planned  to look into this. It becomes quite problematic as you mentioned below because I ifdef the whole rmt module.

Reply:
Well I would not call it "quite problematic". #ifdef-ing only the driver inside `rmt` module is hopefully not a rocket science. Easiest way to do it is what I suggested for the examples:
* Define a `mod driver {}` internal module (not public)
* Move all code _except_ the `peripheral!` peripherals' definitions as well as the RMT peripheral traits (if any) _inside_ the driver module
* Put a `cfg` on top of the `driver `module
* Put a cfg-ed `pub use driver::*` in the `rmt` module

No big deal?

Reply:
No big deal. It's done.